src/Model/Query/GeocodeQuery.php

```diff
@@ -0,0 +1,160 @@
+<?php
+
+/**
+ * This file is part of the Geocoder package.
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ *
+ * @license    MIT License
+ */
+
+namespace Geocoder\Model\Query;
+
+use Geocoder\Exception\InvalidArgument;
+use Geocoder\Model\Bounds;
+use Geocoder\Model\Coordinates;
+use Geocoder\Provider\Provider;
+
+/**
+ * @author Tobias Nyholm <tobias.nyholm@gmail.com>
+ */
+final class GeocodeQuery implements Query
+{
+    /**
+     * @var string
+     */
+    private $text;
+
+    /**
+     * @var Bounds
+     */
+    private $bounds;
+
+    /**
+     * @var string
+     */
+    private $locale;
+
+    /**
+     * @var int
+     */
+    private $limit;
+
+    /**
+     * @var array
+     */
+    private $data;
+
+    /**
+     * @param Coordinates $coordinates
+     */
+    private function __construct($text)
+    {
+        if (empty($text)) {
+            throw new InvalidArgument('Geocode query cannot be empty');
+        }
+
+        $this->text = $text;
+        $this->data = [];
+        $this->limit = Provider::MAX_RESULTS;
+    }
+
+    /**
+     * @param $text
+     *
+     * @return GeocodeQuery
+     */
+    public static function create($text)
+    {
+        return new self($text);
+    }
+
+    /**
+     * @param Bounds $bounds
+     *
+     * @return GeocodeQuery
+     */
+    public function withBounds(Bounds $bounds)
+    {
+        $this->bounds = $bounds;
```

Comment:
I was thinking about making them immutable. I will make the query objects easier to reused and you will not get any unexpected behavior. Though, they are "builder". I compare the GeocodeQuery to Doctrines QueryBuilder. It would not make much sense to have the QueryBuilder immutable, would it? Or maybe Im missing something.